### PR TITLE
[OSE-433] Add the OpenID4VC authorization endpoint

### PR DESCRIPTION
**Pull request by andresuribe87**
# Overview
Adding the authorization endpoint according to https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-authorization-endpoint

# Description
Notably, there are a couple of pieces not included in this PR 

# How Has This Been Tested?
Added unit tests. 

## References
https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-authorization-endpoint

### CODE REVIEW

File: cmd/authserver/main.go [hunk 1]
```diff
@@ -49,8 +50,15 @@ func run() error {
 	shutdown := make(chan os.Signal, 1)
 	signal.Notify(shutdown, os.Interrupt, syscall.SIGTERM)
 
+	// This is the example storage that contains:
+	// * an OAuth2 Client with id "my-client" and secrets "foobar" and "foobaz" capable of all oauth2 and open id connect grant and response types.
+	// * a User for the resource owner password credentials grant type with username "peter" and password "secret".
+	//
+	// You will most likely replace this with your own logic once you set up a real world application.
+	store := storage.NewMemoryStore()
```
Comment (decentralgabe) [inline cmd/authserver/main.go:58]: do we want to swap in bolt here?

File: pkg/authorizationserver/oauth2.go [hunk 2]
```diff
@@ -1,15 +1,52 @@
...
+
+// newSession is a helper function for creating a new session. This may look like a lot of code but since we are
+// setting up multiple strategies it is a bit longer.
+// Usually, you could do:
+//
+//	session = new(fosite.DefaultSession)
+func newSession(user string) *openid.DefaultSession {
+	return &openid.DefaultSession{
+		Claims: &jwt.IDTokenClaims{
+			Issuer:      "https://fosite.my-application.com",
```
Comment (decentralgabe) [inline pkg/authorizationserver/oauth2.go:36]: should replace with some of our stuff? no problem if this is done later

File: pkg/authorizationserver/oauth2.go [hunk 3]
```diff
@@ -1,15 +1,52 @@
...
+			Subject:     user,
+			Audience:    []string{"https://my-client.my-application.com"},
+			ExpiresAt:   time.Now().Add(time.Hour * 6),
```
Comment (decentralgabe) [inline pkg/authorizationserver/oauth2.go:39]: may want a default  or configurable expiry time

File: pkg/authorizationserver/oauth2.go [hunk 4]
```diff
@@ -19,20 +56,73 @@ const (
...
-func NewServer(shutdown chan os.Signal, config *AuthConfig) *Server {
+func NewServer(shutdown chan os.Signal, config *AuthConfig, store *storage.MemoryStore) *Server {
+
+	// This secret is used to sign authorize codes, access and refresh tokens.
+	// It has to be 32-bytes long for HMAC signing. This requirement can be configured via `compose.Config`
+	var secret = make([]byte, 32)
+	_, err := rand.Read(secret)
+	if err != nil {
+		panic(err)
+	}
```
Comment (decentralgabe) [inline pkg/authorizationserver/oauth2.go:67]: one line - and no panic

File: pkg/authorizationserver/oauth2.go [hunk 5]
```diff
@@ -19,20 +56,73 @@ const (
...
 	return &Server{
 		Server: httpServer,
 	}
 }
 
+func loadIssuerMetadata(config *AuthConfig) issuance.IssuerMetadata {
+	// It's ok to panic here as this is only called during startup.
+	jsonData, err := os.ReadFile(config.CredentialIssuerFile)
+	if err != nil {
+		panic(err)
```
Comment (decentralgabe) [inline pkg/authorizationserver/oauth2.go:116]: return err, no panic - here and below

File: pkg/authorizationserver/oauth2_auth.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+func NewAuthService(issuerMetadata *issuance.IssuerMetadata, provider fosite.OAuth2Provider) *AuthService {
+	return &AuthService{issuerMetadata: issuerMetadata, provider: provider}
+}
+
+// AuthEndpoint is a Handler that implements https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-authorization-endpoint
+func (s AuthService) AuthEndpoint(ctx context.Context, rw http.ResponseWriter, req *http.Request) error {
+	ar, err := s.provider.NewAuthorizeRequest(ctx, req)
+	if err != nil {
+		logrus.Errorf("Error occurred in NewAuthorizeRequest: %+v", err)
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply pkg/authorizationserver/oauth2_auth.go:30]: would also add logs to errs below for consistency

File: pkg/authorizationserver/oauth2_auth.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	// locations common data field MUST be set to the Credential Issuer Identifier value
+	if s.issuerMetadata.AuthorizationServer != nil {
+		for i, d := range authorizationDetails {
+			switch d.Type {
+			case "openid_credential":
+				if err := s.processOpenIDCredential(d); err != nil {
+					s.provider.WriteAuthorizeError(ctx, rw, ar, err)
+					return nil
+				}
+				// TODO: support dynamic auth request https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-dynamic-credential-request
```
Comment (decentralgabe) [inline pkg/authorizationserver/oauth2_auth.go:57]: issue tag?

File: pkg/authorizationserver/oauth2_auth.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (andresuribe87) [reply pkg/authorizationserver/oauth2_auth.go]: Done

File: pkg/authorizationserver/oauth2.go (hunk 3 above)
Comment (andresuribe87) [reply pkg/authorizationserver/oauth2.go:40]: Same as above, will evolve in a future PR.

**Comment by andresuribe87** [comment]
@decentralgabe PTAL should have addressed all.

File: cmd/authserver/main.go (hunk 1 above)
Comment (decentralgabe) [reply cmd/authserver/main.go:58]: fine with a todo for now, or task added to the existing oidc issue

**Review by decentralgabe — APPROVED** [review]

**Pull request merged**